[GVN] MemorySSA for GVN: add optional `AllowMemorySSA`

antoniofrighetto opened the pull request:
Preparatory work to migrate from MemoryDependenceAnalysis towards MemorySSA in GVN.

---

Original patch: https://reviews.llvm.org/D118255.
Minor additions in GVNLegacyPass wrt the original version.

alinas commented on llvm/lib/Transforms/Scalar/GVN.cpp:
```diff
@@ -832,7 +838,8 @@ PreservedAnalyses GVNPass::run(Function &F, FunctionAnalysisManager &AM) {
   auto &LI = AM.getResult<LoopAnalysis>(F);
-  auto *MSSA = AM.getCachedResult<MemorySSAAnalysis>(F);
+  auto *MSSA =
+      isMemorySSAEnabled() ? &AM.getResult<MemorySSAAnalysis>(F) : nullptr;
```
Could you add an assert that MemDep and MSSA should never both be enabled?

antoniofrighetto replied:
Added, thanks!

alinas reviewed (approved):
I have some reservations that the flag flip can be done in the short term.
Approving as I think this is work needs to eventually be done.

momchil-velikov committed:
[GVN] MemorySSA for GVN: add optional `AllowMemorySSA`

Preparatory work to migrate from MemoryDependenceAnalysis
towards MemorySSA in GVN.

Co-authored-by: Antonio Frighetto <me@antoniofrighetto.com>

nikic reviewed (commented):
Something here isn't right. There is some compile-time impact and codegen changes: https://llvm-compile-time-tracker.com/compare.php?from=66e41a1a20f2190a800669028a0e80bd86e735ce&to=eb63cd62a4a1907dbd58f12660efd8244e7d81e9&stat=instructions:u

nikic commented on llvm/lib/Transforms/Scalar/GVN.cpp:
```diff
@@ -832,7 +838,10 @@ PreservedAnalyses GVNPass::run(Function &F, FunctionAnalysisManager &AM) {
   auto &LI = AM.getResult<LoopAnalysis>(F);
-  auto *MSSA = AM.getCachedResult<MemorySSAAnalysis>(F);
+  auto *MSSA =
+      isMemorySSAEnabled() ? &AM.getResult<MemorySSAAnalysis>(F) : nullptr;
```
The change here means that we no longer try to update and preserve an already computed MSSA instance. I've reverted this patch for now.
